hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java

```diff
@@ -868,28 +880,40 @@ private InternalPartitionServiceImpl getInternalPartitionService() {
 
     // runs periodically to restart jobs on coordinator failure and perform GC
     private void scanJobs() {
+        long nextScanDelay = maxJobScanPeriodInMillis;
         try {
-            if (!shouldStartJobs()) {
-                return;
-            }
-            Collection<JobRecord> jobs = jobRepository.getJobRecords();
-            for (JobRecord jobRecord : jobs) {
-                JobExecutionRecord jobExecutionRecord = ensureExecutionRecord(jobRecord.getJobId(),
-                        jobRepository.getJobExecutionRecord(jobRecord.getJobId()));
-                startJobIfNotStartedOrCompleted(jobRecord, jobExecutionRecord, "discovered by scanning of JobRecords");
-            }
-            jobRepository.cleanup(nodeEngine);
-            if (!jobsScanned) {
-                synchronized (lock) {
-                    jobsScanned = true;
+            if (isMaster()) {
+                boolean clusterSafe = shouldStartJobs();
+                if (!clusterSafe) {
+                    // use a smaller delay when cluster is not in ready state
+                    nextScanDelay = MIN_JOB_SCAN_PERIOD_MILLIS;
+                } else {
+                    doScanJobs();
                 }
             }
+        } catch (HazelcastInstanceNotActiveException ignored) {
+            // node is shutting down
+            return;
         } catch (Exception e) {
-            if (e instanceof HazelcastInstanceNotActiveException) {
-                return;
-            }
             logger.severe("Scanning jobs failed", e);
         }
+        ExecutionService executionService = nodeEngine.getExecutionService();
+        executionService.schedule(this::scanJobs, nextScanDelay, MILLISECONDS);
+    }
+
+    private void doScanJobs() {
+        Collection<JobRecord> jobs = jobRepository.getJobRecords();
+        for (JobRecord jobRecord : jobs) {
+            JobExecutionRecord jobExecutionRecord = ensureExecutionRecord(jobRecord.getJobId(),
+                    jobRepository.getJobExecutionRecord(jobRecord.getJobId()));
+            startJobIfNotStartedOrCompleted(jobRecord, jobExecutionRecord, "discovered by scanning of JobRecords");
+        }
+        jobRepository.cleanup(nodeEngine);
```

Comment:
I will make a note that we should perhaps move job cleanup to another scheduled task. If some exception would happen above, jobs would not be cleaned up.